src/NumericalAlgorithms/Interpolation/Callbacks/ObserveSurfaceIntegrals.hpp

```diff
@@ -0,0 +1,97 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <string>
+#include <vector>
+
+#include "ApparentHorizons/StrahlkorperGr.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/Dat.hpp"
+#include "IO/H5/File.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace StrahlkorperTags {
+template <typename Frame>
+struct Jacobian;
+template <typename Frame>
+struct NormalOneForm;
+template <typename Frame>
+struct Radius;
+template <typename Frame>
+struct Rhat;
+template <typename Frame>
+struct Strahlkorper;
+}  // namespace StrahlkorperTags
+namespace db {
+template <typename TagsList>
+class DataBox;
+}  // namespace db
+/// \endcond
+
+namespace intrp {
+namespace callbacks {
+
+/// \brief post_interpolation_callback that outputs
+/// surface integrals on a Strahlkorper.
+///
+/// Uses:
+/// - Metavariables
+///   - `temporal_id`
+/// - ConstGlobalCache:
+///   - `FileNameTag`
+/// - DataBox:
+///   - `StrahlkorperTags::items_tags<Frame>`
+///   - `StrahlkorperTags::compute_items_tags<Frame>`
+///   - `TagsToObserve`
+///
+/// This is an InterpolationTargetTag::post_interpolation_callback;
+/// see InterpolationTarget for a description of InterpolationTargetTag.
+template <typename TagsToObserve, typename FileNameTag, typename Frame>
+struct ObserveSurfaceIntegrals {
+  using const_global_cache_tags = tmpl::list<FileNameTag>;
+  template <typename DbTags, typename Metavariables>
+  static void apply(
+      const db::DataBox<DbTags>& box,
+      const Parallel::ConstGlobalCache<Metavariables>& cache,
+      const typename Metavariables::temporal_id& temporal_id) noexcept {
+    // Do the integrals and construct the legend.
+    const auto& strahlkorper = get<StrahlkorperTags::Strahlkorper<Frame>>(box);
+    const auto area_element = StrahlkorperGr::area_element(
+        get<gr::Tags::SpatialMetric<3, Frame>>(box),
+        get<StrahlkorperTags::Jacobian<Frame>>(box),
+        get<StrahlkorperTags::NormalOneForm<Frame>>(box),
+        get<StrahlkorperTags::Radius<Frame>>(box),
+        get<StrahlkorperTags::Rhat<Frame>>(box));
+    std::vector<double> integrals(tmpl::size<TagsToObserve>::value + 1);
+    std::vector<std::string> legend(tmpl::size<TagsToObserve>::value + 1);
+    integrals[0] = temporal_id.value();
+    legend[0] = "Time";
+    size_t s = 1;
+    tmpl::for_each<TagsToObserve>([&](auto tag) noexcept {
+      using Tag = typename decltype(tag)::type;
+      const auto& scalar = get<Tag>(box);
+      integrals[s] = StrahlkorperGr::surface_integral_of_scalar(
+          area_element, scalar, strahlkorper);
+      legend[s] = Tag::name();
+      ++s;
+    });
+
+    // Write to a file.
+    // Do we need file locking?
```

Comment:
This seems like an important question to answer.  If we can't figure out the answer now this should be a ticket, since getting this wrong will cause strange non-reproducible bugs.

Reply:
I agree; that's why I put the comment there.   It depends on usage.  For the accretion disk problem, there will be only one InterpolationTarget/ObserveSurfaceIntegrals ParallelComponent Singleton [for the horizon] that will write to one file, so no locking needed.   If there are multiple InterpolationTarget/ObserveSurfaceIntegrals ParallelComponents for some other problem (wave extraction?), then **if** they all write to different files, again no need for locking.  But if multiple Singletons write to the same file (maybe we want this to cut down the number of files?), then we will need file locking. Unless we switch to parallel hdf5, in which case we presumably would not need locking.